<a href="https://colab.research.google.com/github/sebekpro/python-krakowiak/blob/main/250-data-science-bootcamp-w-jezyku-python/171_180_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 171](#1)
* [Ćwiczenie 172](#2)
* [Ćwiczenie 173](#3)
* [Ćwiczenie 174](#4)
* [Ćwiczenie 175](#5)
* [Ćwiczenie 176](#6)
* [Ćwiczenie 177](#7)
* [Ćwiczenie 178](#8)
* [Ćwiczenie 179](#9)
* [Ćwiczenie 180](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px

np.__version__

'1.26.4'

Uruchom poniższą instrukcję aby pobrać plik _london_bike.csv_ do katalogu roboczego.

In [ ]:
!wget -q https://storage.googleapis.com/esmartdata-courses-files/ds-bootcamp/london_bike.csv

### <a name='1'></a> Ćwiczenie 181
Wczytaj plik _london_bike.csv_ do obiektu DataFrame o nazwie _df_ i wyświetl podstawowe informacje o obiekcie.

In [ ]:
# tutaj wpisz rozwiązanie
df = pd.read_csv('london_bike.csv')


Wyświetl pięć pierwszych wierszy obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie
df.head()

timestamp  cnt   t1   t2    hum  wind_speed  weather_code  \
0  2015-01-04 00:00:00  182  3.0  2.0   93.0         6.0           3.0   
1  2015-01-04 01:00:00  138  3.0  2.5   93.0         5.0           1.0   
2  2015-01-04 02:00:00  134  2.5  2.5   96.5         0.0           1.0   
3  2015-01-04 03:00:00   72  2.0  2.0  100.0         0.0           1.0   
4  2015-01-04 04:00:00   47  2.0  0.0   93.0         6.5           1.0   

   is_holiday  is_weekend  season  
0         0.0         1.0     3.0  
1         0.0         1.0     3.0  
2         0.0         1.0     3.0  
3         0.0         1.0     3.0  
4         0.0         1.0     3.0

Wyświetl pięć ostatnich wierszy obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie
df.tail()

timestamp   cnt   t1   t2   hum  wind_speed  weather_code  \
17409  2017-01-03 19:00:00  1042  5.0  1.0  81.0        19.0           3.0   
17410  2017-01-03 20:00:00   541  5.0  1.0  81.0        21.0           4.0   
17411  2017-01-03 21:00:00   337  5.5  1.5  78.5        24.0           4.0   
17412  2017-01-03 22:00:00   224  5.5  1.5  76.0        23.0           4.0   
17413  2017-01-03 23:00:00   139  5.0  1.0  76.0        22.0           2.0   

       is_holiday  is_weekend  season  
17409         0.0         0.0     3.0  
17410         0.0         0.0     3.0  
17411         0.0         0.0     3.0  
17412         0.0         0.0     3.0  
17413         0.0         0.0     3.0

### <a name='2'></a> Ćwiczenie 182
Zwróć uwagę, że kolumna _timestamp_ jest typu _object_. Zamień ją na typ _datetime_.

__Wskazówka:__ Użyj funkcji _pd.to_datetime()_.

In [ ]:
# tutaj wpisz rozwiązanie
df['timestamp'] = pd.to_datetime(df['timestamp'])
df.head()

timestamp  cnt   t1   t2    hum  wind_speed  weather_code  \
0 2015-01-04 00:00:00  182  3.0  2.0   93.0         6.0           3.0   
1 2015-01-04 01:00:00  138  3.0  2.5   93.0         5.0           1.0   
2 2015-01-04 02:00:00  134  2.5  2.5   96.5         0.0           1.0   
3 2015-01-04 03:00:00   72  2.0  2.0  100.0         0.0           1.0   
4 2015-01-04 04:00:00   47  2.0  0.0   93.0         6.5           1.0   

   is_holiday  is_weekend  season  
0         0.0         1.0     3.0  
1         0.0         1.0     3.0  
2         0.0         1.0     3.0  
3         0.0         1.0     3.0  
4         0.0         1.0     3.0

Do obiektu _df_ dodaj nową kolumnę o nazwie _hour_ i przypisz godzinę z kolumny _timestamp_. Wyświetl kilka początkowych wierszy obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie
df['hour'] = df['timestamp'].dt.hour
df.head()

timestamp  cnt   t1   t2    hum  wind_speed  weather_code  \
0 2015-01-04 00:00:00  182  3.0  2.0   93.0         6.0           3.0   
1 2015-01-04 01:00:00  138  3.0  2.5   93.0         5.0           1.0   
2 2015-01-04 02:00:00  134  2.5  2.5   96.5         0.0           1.0   
3 2015-01-04 03:00:00   72  2.0  2.0  100.0         0.0           1.0   
4 2015-01-04 04:00:00   47  2.0  0.0   93.0         6.5           1.0   

   is_holiday  is_weekend  season  hour  
0         0.0         1.0     3.0     0  
1         0.0         1.0     3.0     1  
2         0.0         1.0     3.0     2  
3         0.0         1.0     3.0     3  
4         0.0         1.0     3.0     4

### <a name='3'></a> Ćwiczenie 183
Do obiektu _df_ dodaj nową kolumnę o nazwie _month_ i przypisz miesiąc z kolumny _timestamp_. Wyświetl kilka początkowych wierszy obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie

df['month'] = df['timestamp'].dt.month
df.head()

timestamp  cnt   t1   t2    hum  wind_speed  weather_code  \
0 2015-01-04 00:00:00  182  3.0  2.0   93.0         6.0           3.0   
1 2015-01-04 01:00:00  138  3.0  2.5   93.0         5.0           1.0   
2 2015-01-04 02:00:00  134  2.5  2.5   96.5         0.0           1.0   
3 2015-01-04 03:00:00   72  2.0  2.0  100.0         0.0           1.0   
4 2015-01-04 04:00:00   47  2.0  0.0   93.0         6.5           1.0   

   is_holiday  is_weekend  season  hour  month  
0         0.0         1.0     3.0     0      1  
1         0.0         1.0     3.0     1      1  
2         0.0         1.0     3.0     2      1  
3         0.0         1.0     3.0     3      1  
4         0.0         1.0     3.0     4      1

Pogrupuj dane po miesiącu (_month_) oraz policz wartość średnią dla zmiennej _hum_. Wynik przypisz do zmiennej _humidity_by_month_ i zresetuj indeks.

In [ ]:
# tutaj wpisz rozwiązanie

humidity_by_month = df.groupby('month')['hum'].mean().reset_index()
humidity_by_month

month        hum
0       1  77.543376
1       2  74.905813
2       3  71.903781
3       4  67.532337
4       5  66.367272
5       6  68.945148
6       7  64.383525
7       8  67.289420
8       9  71.753228
9      10  77.639621
10     11  78.732517
11     12  81.032345

Wykorzystując bibliotekę _plotly express_ zbuduj wykres liniowy średniej wartości zmiennej _hum_ dla danego miesiąca (wykorzystaj obiekt _humidity_by_month_).

__Wskazówka:__ Użyj funkcji _px.line()_.

In [ ]:
# tutaj wpisz rozwiązanie
px.line(humidity_by_month, x = 'month', y = 'hum', title = ' Średnia wartość zmiennej hum dla miesięcy')

Zbuduj ten sam wykres tym razem używając ciemnego szablonu biblioteki _plotly express_.

In [ ]:
# tutaj wpisz rozwiązanie
px.line(humidity_by_month, x = 'month', y = 'hum', title = 'Średnia wartość zmiennej hum dla każdego miesiaca', template = 'plotly_dark')

Odczytaj dla jakiego miesiąca średnia wartość zmiennej _hum_ jest najmniejsza i dla jakiego największa.

### <a name='4'></a> Ćwiczenie 184
Pogrupuj dane na poziomie godziny (_hour_) i następnie wyznacz wartość średnią zmiennej _cnt_. Wynik przypisz do zmiennej _cnt_by_hour_ i zresetuj indeks.

In [ ]:
# tutaj wpisz rozwiązanie
cnt_by_hour = df.groupby('hour')['cnt'].mean().reset_index()
cnt_by_hour

hour          cnt
0      0   290.609116
1      1   200.631215
2      2   136.303745
3      3    94.245492
4      4    73.313454
5      5   110.707351
6      6   466.626722
7      7  1468.745179
8      8  2882.823204
9      9  1653.019257
10    10  1064.660690
11    11  1151.218707
12    12  1433.717421
13    13  1505.739011
14    14  1472.004121
15    15  1564.621399
16    16  1869.801370
17    17  2829.585165
18    18  2629.043956
19    19  1651.510316
20    20  1059.781293
21    21   741.249311
22    22   592.409655
23    23   439.652355

Wykorzystując bibliotekę _plotly express_ zbuduj wykres liniowy średniej wartości zmiennej _cnt_ w rozbiciu na godziny (wykorzystaj obiekt _cnt_by_hour_).

In [ ]:
# tutaj wpisz rozwiązanie
px.line(cnt_by_hour, x = 'hour', y = 'cnt', title = 'Średnia wartość zmiennej cnt w poszczególnych godzinach', labels = {'hour': 'Godzina', 'cnt': 'Średnia cnt'})

W powyższym kodzie:

* Używamy funkcji px.line do wygenerowania wykresu liniowego.
* Parametr x określa zmienną na osi poziomej (godziny).
* Parametr y określa zmienną na osi pionowej (średnia wartość cnt).
* Parametr title ustawia tytuł wykresu.
* Parametr labels pozwala na dostosowanie etykiet osi.

### <a name='5'></a> Ćwiczenie 185
Pogrupuj dane po zmiennych _is_weekend_ i _hour_ i wyznacz średnią wartość zmiennej _cnt_. Wynik przypisz do zmiennej _cnt_by_weekend_hour_ i zresetuj indeks.

In [ ]:
# tutaj wpisz rozwiązanie
cnt_by_weekend_hour = df.groupby(['is_weekend', 'hour'])['cnt'].mean().reset_index()
cnt_by_weekend_hour

is_weekend  hour          cnt
0          0.0     0   215.366795
1          0.0     1   128.001931
2          0.0     2    81.007737
3          0.0     3    53.250000
4          0.0     4    52.248062
5          0.0     5   116.052326
6          0.0     6   604.655106
7          0.0     7  1969.518304
8          0.0     8  3863.922780
9          0.0     9  1997.605769
10         0.0    10  1003.832692
11         0.0    11   968.384615
12         0.0    12  1233.938462
13         0.0    13  1276.556840
14         0.0    14  1211.795761
15         0.0    15  1330.265385
16         0.0    16  1801.692898
17         0.0    17  3232.104046
18         0.0    18  3051.809249
19         0.0    19  1825.682081
20         0.0    20  1133.965318
21         0.0    21   789.131274
22         0.0    22   614.686654
23         0.0    23   434.893617
24         1.0     0   479.810680
25         1.0     1   383.262136
26         1.0     2   276.441176
27         1.0     3   197.434146
28         1.0     4   126.336585
29         1.0     5    97.253659
30         1.0     6   120.555556
31         1.0     7   213.183575
32         1.0     8   415.786408
33         1.0     9   787.391304
34         1.0    10  1218.956098
35         1.0    11  1610.512077
36         1.0    12  1930.775120
37         1.0    13  2074.856459
38         1.0    14  2118.167464
39         1.0    15  2147.708134
40         1.0    16  2039.583732
41         1.0    17  1830.028708
42         1.0    18  1579.210526
43         1.0    19  1216.918269
44         1.0    20   874.677885
45         1.0    21   622.004808
46         1.0    22   537.038462
47         1.0    23   451.653659

W powyższym kodzie:

* Używamy metody groupby, aby pogrupować dane według zmiennych is_weekend oraz hour.
* Następnie dla każdej grupy obliczana jest średnia wartość zmiennej cnt.
*Funkcja reset_index() przekształca wynikowy obiekt Series do DataFrame z domyślnym indeksem.

Wykorzystując bibliotekę _plotly express_ zbuduj wykres liniowy średniej wartości zmiennej _cnt_ w rozbiciu na godziny __w dni poza weekendem__ (wykorzystaj obiekt _cnt_by_weekend_hour_). Nadaj tytuł wykresu: 'Poza weekendem'.

In [ ]:
# tutaj wpisz rozwiązanie

# Filtrowanie danych dla dni poza weekendem (przyjmujemy, że 0 oznacza dni poza weekendem)
cnt_weekdays = cnt_by_weekend_hour[cnt_by_weekend_hour['is_weekend'] == 0]

fig = px.line(cnt_weekdays, x = 'hour', y = 'cnt', title = 'Poza weekendem', labels = {'hour': 'Godzina', 'cnt': 'Średnia cnt'})

fig.show()

* Dane są filtrowane po kolumnie is_weekend.
* Następnie funkcja px.line generuje wykres, gdzie oś X reprezentuje godziny, a oś Y – średnią wartość zmiennej cnt.
* Tytuł wykresu został ustawiony na 'Poza weekendem'.

Wykorzystując bibliotekę _plotly express_ zbuduj wykres liniowy średniej wartości zmiennej _cnt_ w rozbiciu na godziny __w dni weekendowe__ (wykorzystaj obiekt _cnt_by_weekend_hour_). Nadaj tytuł wykresu: 'Weekend'.

In [ ]:
# tutaj wpisz rozwiązanie

# Filtrowanie danych dla dni weekendowych (przyjmujemy, że 1 oznacza weekend)
cnt_weekend = cnt_by_weekend_hour[cnt_by_weekend_hour['is_weekend'] == 1]

# Tworzymy wykres liniowy średniej wartości 'cnt' w zależności od godziny
fig = px.line(cnt_weekend, x = 'hour', y = 'cnt', title = 'Weekend', labels = {'hour': 'Godzina', 'cnt': 'Średnia cnt'})
fig.show()

* Dane są filtrowane po kolumnie is_weekend dla wartości 1 (dni weekendowe).
* Funkcja px.line generuje wykres, gdzie oś X reprezentuje godziny, a oś Y – średnią wartość zmiennej cnt.
* Tytuł wykresu został ustawiony na 'Weekend'.

Czy są jakieś różnice pomiędzy dniami poza weekendem oraz dniami weekendowymi?

### <a name='6'></a> Ćwiczenie 186
Wykorzystując bibliotekę _plotly express_ zbuduj histogram zmiennej _cnt_. Dodaj tytuł wykresu: 'Histogram zmiennej cnt'.

In [ ]:
# tutaj wpisz rozwiązanie

Do histogramu dodaj także wykres pudełkowy zmiennej.

__Wskazówka:__ Użyj parametru _marginal_ funkcji _px.histogram()_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='7'></a> Ćwiczenie 187
Wytnij wiersze dla których zmienna _wind_speed_ jest mniejsza niż 10.0 i zmienna _hum_ jest większa niż 90.0.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='8'></a> Ćwiczenie 188
Wytnij wiersze dla których zmienna _is_weekend_ jest równa 1.0. Następnie skopiuj wynik i przypisz do zmiennej _df_weekend_.


In [ ]:
# tutaj wpisz rozwiązanie

### <a name='9'></a> Ćwiczenie 189
Zapisz obiekt _df_weekend_ do pliku _weekend.xlsx_. Nadaj nazwę arkusza _london_bike_ oraz pomiń zapisanie indeksu do pliku.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='10'></a> Ćwiczenie 190
Zapisz 10 pierwszych rekordów obiektu _df_weekend_ do pliku _head10.html_.

In [ ]:
# tutaj wpisz rozwiązanie